jina/types/document/multimodal.py

```diff
@@ -0,0 +1,80 @@
+from typing import Dict, List, TypeVar
+
+import numpy as np
+
+from . import Document
+from ...proto import jina_pb2
+from ..ndarray.generic import NdArray
+from ...excepts import LengthMismatchException, BadDocType
+
+__all__ = ['MultimodalDocument', 'DocumentContentType']
+
+DocumentContentType = TypeVar('DocumentContentType', bytes, str,
+                              np.ndarray, jina_pb2.NdArrayProto, NdArray)
+
+class MultimodalDocument(Document):
+    """
+    :class:`MultimodalDocument` is a data type created based on Jina primitive data type :class:`Document`.
+
+    It shares the same methods and properties with :class:`Document`, while it focus on modality at chunk level.
+
+    .. warning::
+        - It assumes that every ``chunk`` of a ``document`` belongs to a different modality.
+        - It assumes that every :class:`MultimodalDocument` have at least two chunks.
+    """
+    def __init__(self, document = None, copy: bool = False, **kwargs):
+        """
+
+        :param document: the document to construct from. If ``bytes`` is given
+                then deserialize a :class:`DocumentProto`; ``dict`` is given then
+                parse a :class:`DocumentProto` from it; ``str`` is given, then consider
+                it as a JSON string and parse a :class:`DocumentProto` from it; finally,
+                one can also give `DocumentProto` directly, then depending on the ``copy``,
+                it builds a view or a copy from it.
+        :param copy: when ``document`` is given as a :class:`DocumentProto` object, build a
+                view (i.e. weak reference) from it or a deep copy from it.
+        :param kwargs: other parameters to be set
+        """
+        super().__init__(document=document, copy=copy, **kwargs)
+        self._modality_content_mapping = {}
```

Comment:
Enable the user to create `MultimodalDocument` with the following codes will be awesome
```python
with Document() as c_text:
    c_text.text = 'jina'
    c_text.embedding = np.random.random(2)

with Document() as c_img:
    c_img.blob = np.random.random((28, 28))
    c_img.embedding = np.random.random(4)

d = MultimodalDocument(modality_content={'text': c_text, 'image': c_img})
```

Reply:
yes something like this, and we sill need to take care of the `granularity fields`.

Reply:
@nan-wang @JoanFM I updated the PR with the following interface:
```python
c_text = Document()
c_text.text = 'jina'
c_text.embedding = np.random.random(2)
c_text.modality = 'textual'

c_visual = Document()
c_visual.embedding = np.random.random(2)
c_visual.modality = 'visual'
# Two ways to initialize
# from constructor
d = MultimodalDocument(chunks=[c_textual, c_visual])
# from classmethod
d = MultimodalDocument.from_chunks(chunks=[c_textual, c_visual])
```
The above suggestion poposed by @nan-wang  looks good, while user do not actually need to provide modality information like this: `modality_content={'text': c_text, 'image': c_img}` , since `modality` is a property of chunk, it should be attached to `chunk` directly.

Reply:
I like it because it respects non-mutability of Documents as much as possible, but it could be handy to provide Nan's interface. But let's see what the community demmands regarding this. 

What do u think @nan-wang  it is true that if you offer that interface, u internally need to change those chunks. (What you could do is an interface of the type:

```python
d = MultiModalDocument({'modalitya': contentA, 'modalityB': contentB})
```
And inside it manages the creation of both the chunks and the multimodal document?

Reply:
shall we provide a classmethod called `MultimodalDocument.from_modality_content_mapping` to allow user build the document from a dictionary of embedding/contents?

Reply:
Some fresh thoughts, I think `from_chunkset` would be a better interface compared with `from_chunks` since `ChunkSet` is a primitive data type.

Reply:
Yes, `from_chunkset` sounds better

Reply:
chunk_set is not the right name though, because a `ChunkSet` is a set of `Documents` that are already have a parent. In this case we receive a `docset` and make it a `ChunkSet`.

Reply:
true, we expect `self` as the `reference_doc`, I'll stick with the implementation for now and to see if we can receive some feedback from users.